app/controllers/legacy/sessions_controller.rb

```diff
@@ -0,0 +1,305 @@
+
+    skip_before_action :complete_signup_profile, only: [:destroy]
+
+    before_action :save_new_params_in_session, only: [:start]
+    before_action :store_authorization_url_as_fallback,
+      only: [:start, :create],
+      unless: -> { Settings::Db.store.student_feature_flag }
+
+    before_action :maybe_skip_to_sign_up, only: [:start]
+
+    before_action :allow_iframe_access, only: :reauthenticate
+
+    # If the user arrives to :start already logged in, this means they got linked to
+    # the login page somehow; attempt to redirect to the authorization url stored
+    # earlier
+    before_action :redirect_back, if: -> { signed_in? }, only: :start
+
+    fine_print_skip :general_terms_of_use, :privacy_policy,
+                    only: [:start, :lookup_login, :authenticate, :create, :failure, :destroy, :email_usernames]
+
+    # Login form
+    def start
+      @banners = Banner.active
+    end
+
+    def lookup_login
+      # Most rate limiting happens in CustomIdentity; this separate check needs to be here
+      # here because bad username entries don't even make it all the way to CustomIdentity.
+
+      if too_many_log_in_attempts_by_ip?(ip: request.ip)
+        redirect_to root_url, alert: (I18n.t :"controllers.sessions.too_many_lookup_attempts")
+      else
+        handle_with(SessionsLookupLogin,
+                    success: lambda do
+                      set_login_state(username_or_email: @handler_result.outputs.username_or_email,
+                                      matching_user_ids: @handler_result.outputs.user_ids,
+                                      names: @handler_result.outputs.names,
+                                      providers: @handler_result.outputs.providers.to_hash)
+                      redirect_to :authenticate
+                    end,
+                    failure: lambda do
+                      security_log :login_not_found, tried: @handler_result.outputs.username_or_email
+                      set_login_state(username_or_email: @handler_result.outputs.username_or_email,
+                                      matching_user_ids: @handler_result.outputs.user_ids)
+                      render :start
+                    end)
+      end
+    end
+
+    def reauthenticate
+      handle_with(SessionsReauthenticate,
+                  complete: lambda do
+                    @username_or_email = @handler_result.outputs.username_or_email
+                    @providers = @handler_result.outputs.providers.to_hash
+                  end)
+    end
+
+    def authenticate
+      redirect_to root_path if signed_in?
+    end
+
+    # Handle OAuth callback (actual login)
+    # May add authentication method (OAuth provider) to account
+    def create
+      handle_with(
+        SessionsCreate,
+        user_state: self,
+        pre_auth_state: pre_auth_state,
+        login_providers: get_login_state[:providers],
+        success: -> do
+          authentication = @handler_result.outputs[:authentication]
```

Comment:
consider using a polymorphic object (SessionsCreateUser) to handle this instead of this giant case statement

Reply:
Very good suggestion, actually. I really like that idea, although, there _is_ already a `SessionsCreate` handler that does most of the logic related to this stuff.